### DISCUSSION TITLE
Can I train Whisper in my language ?

### DISCUSSION THREAD

**Discussion by apurvns** [openai/whisper#276] [Q&A] [answered]
I tried using Whisper for my native language Gujarati but result was not good. I am new to AI so want to know can I contribute to this project by  generating training data ?

**Comment by sanchit-gandhi** [comment, marked as answer]
Check-out this blog for fine-tuning Whisper for multilingual ASR with Hugging Face Transformers: https://huggingface.co/blog/fine-tune-whisper

It provides a step-by-step guide to fine-tuning, right from data preparation to evaluation 🤗 There'a Google Colab so you can also run it as a notebook 😉

Feel free to share your training set-up and results, happy to help!

**Reply by DavraYoung** [reply]
First of all, thank you very much for your work! It really helped in understanding.

We are also interested in learning with language/timestamp/nospeech tokens. Are there any examples of how to properly prepare such dataset?

**Reply by giannhskp** [reply]
I am also interested in fine-tuning using timestamps (as mentioned in #587 and my followup question https://github.com/openai/whisper/discussions/587#discussioncomment-4314200). Can you provide us some info on how we can achieve this?

**Reply by sanchit-gandhi** [reply]
Hey @DavraYoung! Super glad to hear the blog post was helpful 🤗 Sounds like a cool task! As for preparing your dataset, you'll need to make sure that it follows the form of predictions that you want the model to make. The model is pre-trained in the following format:
```
<|startoftranscript|> <|lang_id|> <|task_token|> <|notimestamps>| The cat sat on the mat. <|endoftranscript|>
```
Where `<|lang_id|>` corresponds to the language of the audio sample (e.g. `<|es|>` for Spanish) and `<|task_token|>` the task (either `<|transcribe|>` or `<|translate|>`). We add these tokens automatically with the WhisperProcessor in Transformers. If you want to disable them, you can set `tokenizer(batch["sentence"], add_special_tokens=False)` in the `prepare_dataset` function (https://huggingface.co/blog/fine-tune-whisper#prepare-data). You can then format your dataset to follow this token convention and have the model learn the correct lang/task/timestamp tokens according to your dataset annotations

**Reply by sanchit-gandhi** [reply]
Hey @giannhskp! Currently we can only fine-tune for token prediction, but we're looking into adding time-stamp prediction for fine-tuning too! 

Is it crucial that you fine-tune using timestamps? Note that the fine-tuned model will still be capable of predicting time-stamps even if not fine-tuned on this task, since it's pre-trained on a vast amount of time-stamp data and will retain this knowledge

**Reply by giannhskp** [reply]
@sanchit-gandhi Thanks for answering.

By finetuning a whisper model using the format you mentioned above:

``` <|startoftranscript|> <|lang_id|> <|task_token|> <|notimestamps>| The cat sat on the mat. <|endoftranscript|>```
the model loses it's ability to detect timestamps (and split the predicted transcript in smaller chunks).

To be more precise, performing inference on a 5 minute audio, using an original model outputs something like:
```
00:00.000 --> 00:02.000 <random text>

00:02.000 --> 00:04.000 <random text>

00:04.000 --> 00:06.000 <random text>

00:08.000 --> 00:12.000 <random text>

00:12.000 --> 00:16.000 <random text>

...
```
After fine-tuning the model, and performing inference on the same audio, the output always look like:
```
00:00.000 --> 00:30.000 <random text>

00:30.000 --> 01:00.000 <random text>

01:00.000 --> 01:30.000 <random text>

01:30.000 --> 02:00.000 <random text>

...
````
Meaning that the model lost the ability to detect timestamps.

It is crucial for me that the model can retain the ability to split the transcript in small timestamped chucks.

Can i overcome this problem, or it is not possible yet as we can only fine-tune for token prediction (as you said)?

P.S. : The fine-tuning process works great informing the transcribed text, as the model's performance has increased a lot.

**Reply by haely** [reply]
@sanchit-gandhi @DavraYoung Hi, I was looking for NLP datasets in Gujarati for similar goals/projects. What would be a good starting point?

**Reply by DavraYoung** [reply]
@haely I would suggest to check [common voice](https://commonvoice.mozilla.org/) if they have any

**Reply by sanchit-gandhi** [reply]
Hey @giannhskp! Super cool to hear that the fine-tuned model performs well in terms of transcription accuracy, very glad to hear this! It's interesting that the timestamp predictions are degraded so heavily (essentially to random) - this suggests a 'catastrophic forgetting' effect where the model loses all knowledge about timestamp predictions upon fine-tuning with no timestamps. It might be that we have to add some timestamp prediction to the fine-tuning regime to overcome this. You could look to the OpenAI implementation to see how this is done and copy the logic across to the modelling code in Transformers.

Just to clarify your workflow, you're fine-tuning the model in transformers and then converting it back to the OpenAI format to use timestamp prediction? Or are you running everything under transformers? We should have timestamp prediction in transformers very soon! https://github.com/huggingface/transformers/pull/20620

**Reply by sanchit-gandhi** [reply]
Hey @haely and @DavraYoung! Common Voice is indeed a great place to look! CV11 is available on the HF Hub under https://huggingface.co/datasets/mozilla-foundation/common_voice_11_0

**Reply by DavraYoung** [reply]
For any one reading this thread, the transformers timestamp predictions are out. And it seems to work very well on provided pretrained models. 

In my case I need to train whisper to be able to transcribe each word with timestamps in my language:
Here is where I got so far:
Putting timestamps tokens does work during finetuning with transformers:
```
def convert_timestamp_to_token(timestamp, time_precision=20):
    timestamp_begin = whisper["processor"].tokenizer.all_special_ids[-1] + 1
    return timestamp_begin + int(timestamp / time_precision)
```
You can use this function to add timestamps between each word or sentence. But in my case the accuracy is decreasing very much, the loss never gets less than 12% (I assume its because of per word approach)
In my case:
```
for index, caption in enumerate(clip["captions"]):
    is_last = index == len(clip["captions"]) - 1
    start_token = convert_timestamp_to_token(caption["start"])
    end_token = convert_timestamp_to_token(caption["end"])
    text = caption["sentence"].strip()
    tokens += [
        start_token,
        *[
            token
            for token in whisper["processor"]
            .tokenizer(" " + text)
            .input_ids
            if token < end_of_text
        ],
        end_token,
    ]
```


In my case, after one epoch on 400hours dataset my medium model transcribes pretty well the first 7 seconds, but then starts to fail to detect the ending timestamp (I assume its because my dataset has at max 10second audios)

cc:  @sanchit-gandhi 

Has any one tried training whisper to detect word timestamps?

**Reply by sanchit-gandhi** [reply]
Hey @DavraYoung! Thanks for the great write-up! Really cool that the model is able to adapt to per-word timestamps during fine-tuning; this is something that I've been meaning to investigate but haven't gotten round to, so very interesting to see your approach and hear the results from your run! Weird that the loss plateaus - did it keep going down when you fine-tuned without timestamps? Wonder if just the timestamps tokens are the ones that are contributing to the loss, and the word tokens are negligible? Maybe you could have a look at the loss terms on a per-batch basis?

Regarding the audio length, that's quite interesting to hear that the model breaks after 7s. I imagine this is probably down to the distribution of your training data being < 10s. Maybe you could try concatenating audio samples to give longer inputs?

**Reply by Zhuifeng414** [reply]
> Hey @giannhskp! Super cool to hear that the fine-tuned model performs well in terms of transcription accuracy, very glad to hear this! It's interesting that the timestamp predictions are degraded so heavily (essentially to random) - this suggests a 'catastrophic forgetting' effect where the model loses all knowledge about timestamp predictions upon fine-tuning with no timestamps. It might be that we have to add some timestamp prediction to the fine-tuning regime to overcome this. You could look to the OpenAI implementation to see how this is done and copy the logic across to the modelling code in Transformers.
> 
> Just to clarify your workflow, you're fine-tuning the model in transformers and then converting it back to the OpenAI format to use timestamp prediction? Or are you running everything under transformers? We should have timestamp prediction in transformers very soon! [huggingface/transformers#20620](https://github.com/huggingface/transformers/pull/20620)

Same issue when I fine-tuned the Whisper model on a samll dataset without timestamps; the model displayed 'catastrophic forgetting'. This led to worse performance on samples beyond the fine-tuning dataset. Do you have any solution to solve it?

**Reply by sanchit-gandhi** [reply]
PEFT is more resilient to catastrophic forgetting and gives comparable performance: https://github.com/Vaibhavs10/fast-whisper-finetuning

**Reply by sutungpo** [reply]
> Check-out this blog for fine-tuning Whisper for multilingual ASR with Hugging Face Transformers: https://huggingface.co/blog/fine-tune-whisper
> 
> It provides a step-by-step guide to fine-tuning, right from data preparation to evaluation 🤗 There'a Google Colab so you can also run it as a notebook 😉
> 
> Feel free to share your training set-up and results, happy to help!
@sanchit-gandhi 
For fine-tuning Whisper translation, the evaluation metric should be the Word Error Rate (WER) metric, as mentioned in your article, or other metrics such as BLEU?
